librosa/segment.py

```diff
@@ -38,13 +39,213 @@
+        `exp( - distance(i, j) / bandwidth)` where `bandwidth` is
+        as specified below.
+
+    bandwidth : None or float > 0
+        If using ``mode='affinity'``, this can be used to set the
+        bandwidth on the affinity kernel.
+
+        If no value is provided, it is set automatically to the median
+        distance between furthest nearest neighbors.
+
+    Returns
+    -------
+    rec : np.ndarray or scipy.sparse.csr_matrix, [shape=(n_times1, n_times2)]
+        Recurrence matrix
+
+    See Also
+    --------
+    rec
+    sklearn.neighbors.NearestNeighbors
+    scipy.spatial.distance.cdist
+    librosa.feature.stack_memory
+    recurrence_to_lag
+
+    Notes
+    -----
+    This function caches at level 30.
+
+    Examples
+    --------
+    Find nearest neighbors in MFCC space
+
+    >>> y, sr = librosa.load(librosa.util.example_audio_file())
+    >>> mfcc = librosa.feature.mfcc(y=y, sr=sr)
+    >>> R = librosa.segment.recurrence_matrix(mfcc)
+
+    Or fix the number of nearest neighbors to 5
+
+    >>> R = librosa.segment.recurrence_matrix(mfcc, k=5)
+
+    Suppress neighbors within +- 7 samples
+
+    >>> R = librosa.segment.recurrence_matrix(mfcc, width=7)
+
+    Use cosine similarity instead of Euclidean distance
+
+    >>> R = librosa.segment.recurrence_matrix(mfcc, metric='cosine')
+
+    Require mutual nearest neighbors
+
+    >>> R = librosa.segment.recurrence_matrix(mfcc, sym=True)
+
+    Use an affinity matrix instead of binary connectivity
+
+    >>> R_aff = librosa.segment.recurrence_matrix(mfcc, mode='affinity')
+
+    Plot the feature and recurrence matrices
+
+    >>> import matplotlib.pyplot as plt
+    >>> plt.figure(figsize=(8, 4))
+    >>> plt.subplot(1, 2, 1)
+    >>> librosa.display.specshow(R, x_axis='time', y_axis='time')
+    >>> plt.title('Binary recurrence (symmetric)')
+    >>> plt.subplot(1, 2, 2)
+    >>> librosa.display.specshow(R_aff, x_axis='time', y_axis='time',
+    ...                          cmap='magma_r')
+    >>> plt.title('Affinity recurrence')
+    >>> plt.tight_layout()
+
+    '''
+    data1 = np.atleast_2d(data1)
+    data2 = np.atleast_2d(data2)
+
+    if data1.shape[0] != data2.shape[0]:
+        raise ValueError("data1 and data2 must have the same first dimension")
+
+    data1 = np.swapaxes(data1, -1, 0)
+    n_times1 = data1.shape[0]
+    data1 = data1.reshape((n_times1, -1))
+
+    data2 = np.swapaxes(data2, -1, 0)
+    n_times2 = data2.shape[0]
+    data2 = data2.reshape((n_times2, -1))
+
+    n_feat = data1.shape[-1]
+
+    if width < 1:
+        raise ParameterError('width must be at least 1')
+
+    if mode not in ['connectivity', 'distance', 'affinity']:
+        raise ParameterError(("Invalid mode='{}'. Must be one of "
+                              "['connectivity', 'distance', "
+                              "'affinity']").format(mode))
+    if k is None:
+        if n_times1 > 2 * width + 1:
+            k = 2 * np.ceil(np.sqrt(n_feat - 2 * width + 1))
```

Comment:
It would be good to describe what this equation does.
The naming choice `k` is a bit confusing because the number of neighbors is not `k` but `min(n_times1-1, k + 2 * width)`